src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java

```diff
@@ -491,13 +491,13 @@ public void visitToken(DetailAST ast) {
      * Checks if current method's return type or variable's type is verifiable
      * according to <b>memberModifiers</b> option.
      *
-     * @param methodOrVariableDef METHOD_DEF or VARIABLE_DEF ast node.
+     * @param ast METHOD_DEF or VARIABLE_DEF ast node.
      * @return true if member is verifiable according to <b>memberModifiers</b> option.
      */
-    private boolean isVerifiable(DetailAST methodOrVariableDef) {
+    private boolean isVerifiable(DetailAST ast) {
         boolean result = true;
-        if (!memberModifiers.isEmpty()) {
-            final DetailAST modifiersAst = methodOrVariableDef
+        if (!memberModifiers.isEmpty() && ast.getType() != TokenTypes.RECORD_COMPONENT_DEF) {
```

Comment:
It is not clear why modifier is null.
We have modifiers on record https://github.com/checkstyle/checkstyle/blob/988cbe38621df6cf079c975fc9ca87b45a3c4af9/src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java#L5447

Records should have modifiers same as other types.

Reply:
You mean I need to change `TokenTypes.RECORD_COMPONENT_DEF -> TokenTypes.RECORD_DEF` ?

Reply:
@romani sir, can you please clarify what exactly I need to change?

Reply:
@romani  sir, can you please clarify what exactly I need to change?

Reply:
I do not understand why we simply skip validation
https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java#L5513
It can have illegal type.

Please create input with violation on record component.

Reply:
@romani sir,
After changing 
public record InputIllegalTypeRecordsWithMemberModifiersSpecified
        (UUID productId, String identifier) 
-->
public record InputIllegalTypeRecordsWithMemberModifiersSpecified
        (HashSet productId, String identifier) {

This is failing and It says Violation lines for   (HashSet productId, String identifier) {expected to be empty.

Reply:
This still worries me.
I probably need to debug this code to understand it

Reply:
please make code like:

```
            case TokenTypes.RECORD_COMPONENT_DEF:
                visitComponentDef(ast);
                break;
```

```
visitComponentDef() {
    if (isContainVerifiableType(variableDef)) {
            checkClassName(variableDef);
        }
}
```

Reply:
@romani condition in `visitComponentDef` if statement should be extended to check `memberModifiers` for `private` and `final` modifiers. The two following type declarations should be equivalent to this check, and we should expect violations on both with default config (how it works currently) and with any combination of `FINAL` and/or `LITERAL_PRIVATE`:

```java
import java.util.TreeSet;

class TestClass {
    private final TreeSet x;

    TestClass(TreeSet x) {
        this.x = x;
    }
}

record TestRecord(TreeSet x) {}

```

Reply:
@romani @nmancus1
After adding Romani sir's code:

```
case TokenTypes.RECORD_COMPONENT_DEF:
                visitComponentDef(ast);
                break;
```

```
  /**
     * Checks type of ComponentDef.
     *
     * @param variableDef to check.
     */
    private void visitComponentDef(DetailAST variableDef) {
        if (isContainVerifiableType(variableDef)) {
            checkClassName(variableDef);
        }
    }
```
and changing the access specifier from `PUBLIC RECORD`  to `PRIVATE FINAL RECORD`

I found no violation for HashSet so I removed  `// violation`

Reply:
After Adding Romani's code:

```
[ERROR] Failures: 
[ERROR]   IllegalTypeCheckTest.testArrayTypes:194->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeArrays.java differ.
missing (4): 20, 22, 25, 30
---
expected   : [20, 22, 24, 25, 29, 30]
but was    : [24, 29]
[ERROR]   IllegalTypeCheckTest.testGenerics:232->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeTestGenerics.java differ.
missing (5): 28, 29 [2 copies], 33, 68
---
expected   : [28, 29, 29, 32, 33, 33, 36, 36, 38, 40, 40, 42, 42, 46, 46, 55, 68]
but was    : [32, 33, 36, 36, 38, 40, 40, 42, 42, 46, 46, 55]
[ERROR]   IllegalTypeCheckTest.testIgnoreMethodNames:122->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeTestIgnoreMethodNames.java differ.
missing (2): 23, 26
---
expected   : [23, 26, 34, 43, 60, 62]
but was    : [34, 43, 60, 62]
[ERROR]   IllegalTypeCheckTest.testIllegalTypeEnhancedInstanceof:327->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources-noncompilable\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeTestEnhancedInstanceof.java differ.
missing (4): 28, 31, 35, 41
---
expected   : [28, 31, 35, 40, 41]
but was    : [40]
[ERROR]   IllegalTypeCheckTest.testIllegalTypeNewArrayStructure:356->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeNewArrayStructure.java differ.
missing (1): 26
---
expected   : [26]
but was    : []
[ERROR]   IllegalTypeCheckTest.testIllegalTypeRecordsAndCompactCtors:344->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources-noncompilable\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeRecordsAndCompactCtors.java differ.
missing (5): 27, 32, 35, 39, 48
---
expected   : [27, 31, 32, 35, 39, 40, 48]
but was    : [31, 40]
[ERROR]   IllegalTypeCheckTest.testLegalAbstractClassNames:153->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeTestLegalAbstractClassNames.java differ.
missing (1): 26
---
expected   : [26, 34, 35, 60, 62]
but was    : [34, 35, 60, 62]
[ERROR]   IllegalTypeCheckTest.testMemberModifiers:288->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeTestMemberModifiers.java differ.
missing (4): 22, 25, 39, 49
---
expected   : [22, 25, 32, 33, 39, 41, 49]
but was    : [32, 33, 41]
[ERROR]   IllegalTypeCheckTest.testPlainAndArrayTypes:207->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeTestPlainAndArraysTypes.java differ.
missing (3): 20, 24, 36
---
expected   : [20, 24, 26, 35, 36]
but was    : [26, 35]
[ERROR]   IllegalTypeCheckTest.testSameFileNameFalsePositive:164->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeSameFileNameFalsePositive.java differ.
missing (2): 28, 43
---
expected   : [28, 43]
but was    : []
[ERROR]   IllegalTypeCheckTest.testSameFileNameGeneral:180->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeTestSameFileNameGeneral.java differ.
missing (6): 25, 39, 40, 42, 43, 44
---
expected   : [25, 29, 31, 39, 40, 42, 43, 44]
but was    : [29, 31]
[ERROR]   IllegalTypeCheckTest.testStarImports:260->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeTestStarImports.java differ.
missing (1): 25
---
expected   : [25]
but was    : []
[ERROR]   IllegalTypeCheckTest.testStaticImports:272->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeTestStaticImports.java differ.
missing (1): 28
---
expected   : [28, 30]
but was    : [30]
[ERROR]   IllegalTypeCheckTest.testValidateAbstractClassNamesSetToTrue:52->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeTestAbstractClassNamesTrue.java differ.
missing (1): 44
---
expected   : [27, 44, 46, 50]
but was    : [27, 46, 50]
[INFO] 
[ERROR] Tests run: 3652, Failures: 14, Errors: 0, Skipped: 1
```

Reply:
Should I resolve all these error or stick to only the current issue?

Reply:
Update to check should look something like this: https://github.com/checkstyle/checkstyle/commit/959d778801a28d1e40d3f1bc268ec820660b1df0

Reply:
After doing this, there will be no violations, right?

Reply:
@nmancus1 After doing like #959d778 I am getting no violations for HashSet, is it okay or are we expecting any violation?

Reply:
Amend all test cases to have comparable class definition for each given record definition to show identical violations (or lack thereof) as detailed in https://github.com/checkstyle/checkstyle/pull/10958/files#r758611207 . Also, please read about record definitions and record components to better understand how they relate to class definitions athttps://openjdk.java.net/jeps/359.

Reply:
@nmancus1  Please check my changes and Kindly reply for the same:
https://gist.github.com/Rahulkhinchi03/73db12098dc8d808b750ab3bc7d3c76e

Reply:
You should amend test cases to include an equivalent class definition. This will help you to convince reviewers (and yourself) that:

```java

class TestClass {
    private final TreeSet x; // <- this declaration

    TestClass(TreeSet x) {
        this.x = x;
    }
}

```

is treated the same, with any configuration, that:

```java

record TestRecord(
        TreeSet x // <- this record component is
) {}

```

So for member modifiers like `LITERAL_PUBLIC, LITERAL_PROTECTED, LITERAL_STATIC`, the above examples should produce no violations, since `TestClass.x` is `private final` and `TestRecord.x` has implied modifiers `final` and `private`. The above examples would produce violations for the default config, and should for member modifiers like `FINAL` and `LITERAL_PRIVATE` or any combination thereof.

Reply:
Thank you so much @nmancus1  for the easy explanation.
@nmancus1 Please help me with the coverage issue: branch coverage is 0.97%

Reply:
Also, I am getting an error of `javac14` that I cannot use` private` in the private final record

Reply:
To start with, you should have two or three test input files that are the same code, with different configs (this will be necessary for coverage), to show that violations (or lack thereof) are the same for test input that **includes** the following:

```java

import java.util.TreeSet;

class TestClass {
    private final TreeSet x; // <- any 'violation' or 'ok' here

    TestClass(TreeSet x) {
        this.x = x;
    }
}

record TestRecord(
        TreeSet x // // <- should be identical to `violation' or 'ok' here, for any config
) {}


```

Reply:
I am sorry but It is quite complex for be as a beginner

Reply:
See https://github.com/nmancus1/checkstyle/commit/a50a96ffba8b8d8a0fa942bf4f1a7399b636a2f5

Reply:
Thanks a lot! @nmancus1 I understood now where my mistake was.

Reply:
branches covered ratio is 0.99, but expected minimum is 1.00

Reply:
I am still getting this issue @nmancus1 
`branches covered ratio is 0.99, but expected minimum is 1.00
`

Even after changing everything

Reply:
Push your changes so that I can help; note that you do not need to modify the `isVerifiable` method in any way now, this method does not deal with record components at all anymore.

Reply:
> 

Using this code, I am getting violation error in file ` InputIllegalTypeRecordsAndCompactCtor`,

1. Should I use this code only and resolve these errors too?
2. Or Should I use: 
 ```
private void visitRecordComponentDef(DetailAST recordComponentDef) {
        if (memberModifiers.isEmpty()) {
            checkClassName(recordComponentDef);
        }
    }
```

Using point 2, I am getting no error.

Reply:
@nmancus1 I am again getting the branch and line overage issue. I only updated the Java Doc.

Reply:
@nmancus1  I am again getting the branch and line overage issue. I only updated the Java Doc and Added Romani sir's Code so I was getting an error and I updated to https://github.com/checkstyle/checkstyle/pull/10958#discussion_r762512351 

```
    /**
     * Checks a record component definition. Record components are private
     * final fields, so we do not consider memberModifiers field.
     *
     * @param recordComponentDef the record component definition to check.
     */
    private void visitRecordComponentDef(DetailAST recordComponentDef) {
            if (isContainVerifiableType(recordComponentDef)) {
                checkClassName(recordComponentDef);
            }
    }
```

To be precise:  `checkClassName(recordComponentDef);` This line is red and not in coverage